File: src/test/java/net/dontdrinkandroot/gitki/test/AbstractWicketTest.java
package net.dontdrinkandroot.gitki.test;

import net.dontdrinkandroot.gitki.model.User;
import net.dontdrinkandroot.gitki.wicket.GitkiWebApplication;
import net.dontdrinkandroot.gitki.wicket.GitkiWebSession;
import net.dontdrinkandroot.gitki.wicket.page.SignInPage;
import org.apache.commons.lang3.reflect.ConstructorUtils;
import org.apache.wicket.Page;
import org.apache.wicket.authorization.UnauthorizedInstantiationException;
import org.apache.wicket.request.mapper.parameter.PageParameters;
import org.apache.wicket.util.tester.WicketTester;
import org.junit.jupiter.api.Assertions;
import org.junit.jupiter.api.BeforeEach;
import org.springframework.beans.factory.annotation.Autowired;

import java.lang.reflect.InvocationTargetException;

/**
 * @author Philip Washington Sorst <philip@sorst.net>
 */
public abstract class AbstractWicketTest extends AbstractIntegrationTest
{
    @Autowired
    protected GitkiWebApplication wicketApplication;

    private WicketTester wicketTester;

    @BeforeEach
    public void before()
    {
        this.wicketTester = new WicketTester(this.wicketApplication);
    }

    public WicketTester getWicketTester()
    {
        return this.wicketTester;
    }

    protected void assertLoginRequired(Class<? extends Page> pageClass)
    {
        this.assertLoginRequired(pageClass, new PageParameters());
    }

    protected void assertLoginRequired(Class<? extends Page> pageClass, PageParameters parameters)
    {
        this.wicketTester.startPage(pageClass, parameters);
        this.wicketTester.assertRenderedPage(SignInPage.class);
    }

    protected <P extends Page> P assertPageAccessible(Class<P> pageClass, User user, Object... parameters)
    {
        GitkiWebSession.get().invalidate();
        if (null != user) {
            GitkiWebSession.get().signIn(user);
        }
        P page = this.startPage(pageClass, parameters);
        this.wicketTester.assertRenderedPage(pageClass);

        return page;
    }

    protected void assertPageInaccessible(Class<? extends Page> pageClass, User user, Object... parameters)
    {
        GitkiWebSession.get().invalidate();
        if (null != user) {
            GitkiWebSession.get().signIn(user);
        }
        try {
            this.startPage(pageClass, parameters);
            Assertions.fail("UnauthorizedInstantiationException expected");
        } catch (UnauthorizedInstantiationException e) {
            /* Expected */
        }
    }

    protected <C extends Page> C startPage(Class<C> pageClass, Object[] parameters)
    {
        if (null == parameters || 0 == parameters.length) {
            return this.wicketTester.startPage(pageClass);
        }

        if (1 == parameters.length && parameters[0] instanceof PageParameters) {
            return this.wicketTester.startPage(pageClass, (PageParameters) parameters[0]);
        }

        try {
            C page = ConstructorUtils.invokeConstructor(pageClass, parameters);
            return this.wicketTester.startPage(page);
        } catch (NoSuchMethodException | InstantiationException | InvocationTargetException | IllegalAccessException e) {
            if (e instanceof InvocationTargetException) {
                InvocationTargetException invocationTargetException = (InvocationTargetException) e;
                if (invocationTargetException.getTargetException() instanceof UnauthorizedInstantiationException) {
                    throw (UnauthorizedInstantiationException) invocationTargetException.getTargetException();
                }
            }
            throw new RuntimeException(e);
        }
    }
}

File: src/test/kotlin/net/dontdrinkandroot/gitki/test/AbstractWicketTest.kt
package net.dontdrinkandroot.gitki.test

import net.dontdrinkandroot.gitki.model.User
import net.dontdrinkandroot.gitki.wicket.GitkiWebApplication
import net.dontdrinkandroot.gitki.wicket.getGitkiSession
import net.dontdrinkandroot.gitki.wicket.page.SignInPage
import org.apache.commons.lang3.reflect.ConstructorUtils
import org.apache.wicket.Page
import org.apache.wicket.authorization.UnauthorizedInstantiationException
import org.apache.wicket.request.mapper.parameter.PageParameters
import org.apache.wicket.util.tester.WicketTester
import org.junit.jupiter.api.Assertions
import org.junit.jupiter.api.BeforeEach
import org.springframework.beans.factory.annotation.Autowired
import java.lang.reflect.InvocationTargetException

abstract class AbstractWicketTest : AbstractIntegrationTest() {

    @Autowired
    protected lateinit var wicketApplication: GitkiWebApplication

    lateinit var wicketTester: WicketTester

    @BeforeEach
    fun before() {
        wicketTester = WicketTester(wicketApplication)
    }

    protected fun assertLoginRequired(pageClass: Class<out Page?>, parameters: PageParameters? = PageParameters()) {
        wicketTester.startPage(pageClass, parameters)
        wicketTester.assertRenderedPage(SignInPage::class.java)
    }

    protected fun <P : Page?> assertPageAccessible(pageClass: Class<P>, user: User?, vararg parameters: Any?): P {
        getGitkiSession().invalidate()
        if (null != user) {
            getGitkiSession().signIn(user)
        }
        val page = startPage(pageClass, parameters)
        wicketTester.assertRenderedPage(pageClass)
        return page
    }

    protected fun assertPageInaccessible(pageClass: Class<out Page>, user: User?, vararg parameters: Any?) {
        getGitkiSession().invalidate()
        if (null != user) {
            getGitkiSession().signIn(user)
        }
        try {
            startPage(pageClass, parameters)
            Assertions.fail<Any>("UnauthorizedInstantiationException expected")
        } catch (e: UnauthorizedInstantiationException) {
            /* Expected */
        }
    }

    protected fun <C : Page?> startPage(pageClass: Class<C>, parameters: Array<out Any?>): C {
        if (parameters.isEmpty()) {
            return wicketTester.startPage(pageClass)
        }
        return if (1 == parameters.size && parameters[0] is PageParameters) {
            wicketTester.startPage(pageClass, parameters[0] as PageParameters?)
        } else try {
            val page = ConstructorUtils.invokeConstructor(pageClass, *parameters)
            wicketTester.startPage(page)
        } catch (e: NoSuchMethodException) {
            if (e is InvocationTargetException) {
                if (e.targetException is UnauthorizedInstantiationException) {
                    throw (e.targetException as UnauthorizedInstantiationException)
                }
            }
            throw RuntimeException(e)
        } catch (e: InstantiationException) {
            if (e is InvocationTargetException) {
                if (e.targetException is UnauthorizedInstantiationException) {
                    throw (e.targetException as UnauthorizedInstantiationException)
                }
            }
            throw RuntimeException(e)
        } catch (e: InvocationTargetException) {
            if (e.targetException is UnauthorizedInstantiationException) {
                throw (e.targetException as UnauthorizedInstantiationException)
            }
            throw RuntimeException(e)
        } catch (e: IllegalAccessException) {
            if (e is InvocationTargetException) {
                if (e.targetException is UnauthorizedInstantiationException) {
                    throw (e.targetException as UnauthorizedInstantiationException)
                }
            }
            throw RuntimeException(e)
        }
    }
}